compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/inference/ConstraintSystemBuilderImpl.kt

```diff
@@ -64,7 +65,7 @@ open class ConstraintSystemBuilderImpl(private val mode: Mode = ConstraintSystem
         EQUAL(EXACT_BOUND)
     }
 
-    internal val allTypeParameterBounds = LinkedHashMap<TypeVariable, TypeBoundsImpl>()
+    internal val allTypeParameterBounds = SmartIdentityTable<TypeVariable, TypeBoundsImpl>()
```

Comment:
Unfortunately we'll now lose the order of iteration reflecting the sequence in which the type variables were added, as soon as the size of this map becomes greater than 10, which I think could lead to very hard-to-debug problems (those in which you can't minimize the test case because it starts to behave correctly)